### PR TITLE
fix: adam(w) ignore stride mismatch when dim is size 1

### LINKED ISSUE #106951: stride of gradient is not same as the corresponding tensor
### 🐛 Describe the bug

When I tried to use torch.optimizer.Adam with fused=True, I got the following error:

```text
  File "/home/weixu/venvs/working/lib/python3.10/site-packages/torch/optim/adam.py", line 591, in _fused_adam
    torch._fused_adam_(
RuntimeError: params, grads, exp_avgs, and exp_avg_sqs must have same dtype, device, and layout
```

With some debug, I found this is caused by the mismatching of the stride of a parameter and its gradient. Indeed, as the following code shows, the strides do mismatch in some corner case:

```python
import torch
d = 1
device = 'cuda'
w = torch.nn.Parameter(torch.zeros((2, d, 100), device=device))
x = torch.rand((10, 2, 100), device=device)
y = torch.bmm(x.transpose(0, 1), w.transpose(1,2))
y.sum().backward()
print(w.stride(), w.grad.stride())
assert w.stride() == w.grad.stride()
```
It still fails if bmm is replaced by `torch.baddbmm` or `torch.einsum()`. Fails with device='cpu' too.
 `w.grad.stride()` is `(100, 1, 1)`, while `w.stride()` is `(100, 100, 1)`

The assert can pass if d > 1.

Tested on botch torch 1.13.0+cu116  and torch2.0.1+cu118

### Versions

Collecting environment information...
PyTorch version: 1.13.0+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.19.0-46-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 470.199.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          32
On-line CPU(s) list:             0-31
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Core(TM) i9-7960X CPU @ 2.80GHz
CPU family:                      6
Model:                           85
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        4
CPU max MHz:                     4600.0000
CPU min MHz:                     1200.0000
BogoMIPS:                        5599.85
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single pti ssbd mba ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req md_clear flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       512 KiB (16 instances)
L1i cache:                       512 KiB (16 instances)
L2 cache:                        16 MiB (16 instances)
L3 cache:                        22 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-31
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:               Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Meltdown:          Mitigation; PTI
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Mitigation; IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; Clear CPU buffers; SMT vulnerable

Versions of relevant libraries:
[pip3] numpy==1.23.0
[pip3] numpy-quaternion==2022.4.3
[pip3] pytorch3d==0.7.3
[pip3] torch==1.13.0+cu116
[pip3] torchaudio==0.13.0+cu116
[pip3] torchvision==0.14.0+cu116
[pip3] triton==2.0.0
[conda] Could not collect

### PR DESCRIPTION
**Pull request by functionstackx**
Fixes #106951

when dim is of size 1, stride mismatch does not matter

As discussed in #78050 strides are inconsistent and may sometimes change when dim 1 during autograd.

Thus we fix the fused_adam(w) check to account for this case.

cc: @clchan @ezhang887

cc for review @janeyx99 @albanD 

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### CODE REVIEW

File: aten/src/ATen/native/ForeachUtils.h [hunk 1]
```diff
@@ -134,7 +134,18 @@ bool check_fast_path_restrictions(
         return false;
       }
       if (tensorLists[0][j].strides() != tensor_list[j].strides()) {
-        return false;
```
Comment (janeyx99) [inline aten/src/ATen/native/ForeachUtils.h:137]: Should probably verify that the length of tensor_list[j].strides is the same as tensorLists[0][j].strides()

File: aten/src/ATen/native/ForeachUtils.h [hunk 2]
```diff
@@ -134,7 +134,18 @@ bool check_fast_path_restrictions(
...
+        for (auto it1 = tensorLists[0][j].sizes().begin(),
```
Comment (janeyx99) [inline aten/src/ATen/native/ForeachUtils.h:137]: Can we rename it1, it2, and it3 to something that would help the reader more, maybe (feel free to adapt) ref_size, ref_stride, tensor_list_stride?

**Review by janeyx99 — COMMENTED** [review]
Thanks for the change! Can you also add unit tests in test_foreach.py? e.g., testing that the stride can be different when the size of a dim is 1.

also cc @crcrpar

File: aten/src/ATen/native/ForeachUtils.h (hunk 1 above)
Comment (OrenLeung) [reply aten/src/ATen/native/ForeachUtils.h:137]: addressed in b50604ae7ad86746350d16a8dc2d0f683d71f6cf

File: aten/src/ATen/native/ForeachUtils.h (hunk 2 above)
Comment (OrenLeung) [reply aten/src/ATen/native/ForeachUtils.h]: addressed in b50604ae7ad86746350d16a8dc2d0f683d71f6cf

**Comment by OrenLeung** [comment]
thanks for the review @janeyx99 i added yr suggestions

**Comment by crcrpar** [comment]
I don't think so as `can_use_fast_route` calling `check_fast_path_restrictions` is called in many places  https://github.com/search?q=repo%3Apytorch%2Fpytorch%20can_use_fast_route&type=code

**Comment by OrenLeung** [comment]
> It seems to have some effects on foreach functions as well. Wouldn't it make sense to add some test cases to foreach ones as well?

ahhh i see it now. thank you for pointing that out. will add tests for this

File: aten/src/ATen/native/ForeachUtils.h [hunk 3]
```diff
@@ -130,11 +130,32 @@ bool check_fast_path_restrictions(
   // strides.
   for (const auto& tensor_list : tensorLists) {
     for (const auto j : c10::irange(tensorLists[0].size())) {
-      if (tensorLists[0][j].sizes() != tensor_list[j].sizes()) {
+      const auto& ref_size = tensorLists[0][j].sizes();
+      const auto& ref_stride = tensorLists[0][j].strides();
+      const auto& tensor_list_size = tensor_list[j].sizes();
+      const auto& tensor_list_stride = tensor_list[j].strides();
```
Comment (janeyx99) [inline aten/src/ATen/native/ForeachUtils.h:136]:
```suggestion
      const auto& ref_sizes = tensorLists[0][j].sizes();
      const auto& ref_strides = tensorLists[0][j].strides();
      const auto& tensor_list_sizes = tensor_list[j].sizes();
      const auto& tensor_list_strides = tensor_list[j].strides();
```

nit for clarity

File: test/test_foreach.py [hunk 4]
```diff
@@ -1114,6 +1115,16 @@ def test_0dim_tensor_overload_exception(self):
...
+    @onlyCUDA
+    def test_ignore_stride_mismatch_when_dim_is_0(self):
+        w = torch.nn.Parameter(torch.zeros((2, 1, 100), device='cuda'))
+        x = torch.rand((10, 2, 100), device='cuda')
+        optimizer = optim.AdamW([w], lr=0.0001, fused=True)
+        y = torch.bmm(x.transpose(0, 1), w.transpose(1, 2))
+        y.sum().backward()
+        self.assertNotEqual(w.stride(), w.grad.stride())
+        optimizer.step()
+
```
Comment (janeyx99) [inline test/test_foreach.py:1127]: Would be good to also add a test where we expect the error if the stride is mismatched and the size is not 1 to make sure we did not relax the check too much

**Comment by janeyx99** [comment]
@OrenLeung do you have an ETA for this change?

**Pull request closed without merging**